mlir/lib/Analysis/Presburger/PresburgerBasicSet.cpp

```diff
@@ -492,3 +503,311 @@ void PresburgerBasicSet::dumpISL() const {
+    SmallVector<int64_t, 8> coeffs = copyWithNonRedundant(
+        nonRedundantExist, nonRedundantDivs, eq.getCoeffs());
+
+    newEqs.push_back(EqualityConstraint(coeffs));
+  }
+
+  unsigned variableOffset = nDim + nParam + nonRedundantExist.size();
+  for (unsigned i : nonRedundantDivs) {
+    DivisionConstraint div = divs[i - divOffset];
+
+    SmallVector<int64_t, 8> coeffs = copyWithNonRedundant(
+        nonRedundantExist, nonRedundantDivs, div.getCoeffs());
+
+    newDivs.push_back(DivisionConstraint(coeffs, div.getDenominator(),
+                                         variableOffset + newDivs.size()));
+  }
+
+  // Assign new vectors
+  ineqs = newIneqs;
+  eqs = newEqs;
+  divs = newDivs;
+
+  // Change dimensions
+  nExist = nonRedundantExist.size();
+}
+
+void PresburgerBasicSet::alignDivs(PresburgerBasicSet &bs1,
+                                          PresburgerBasicSet &bs2) {
+  // Assert that bs1 has more divisions than bs2
+  if (bs1.getNumDivs() < bs2.getNumDivs()) {
+    alignDivs(bs2, bs1);
+    return;
+  }
+
+  bs1.simplify();
+  bs2.simplify();
+
+  // TODO: Is there a better stratergy than this ?
+  // Append extra existentials
+  if (bs1.nExist > bs2.nExist) {
+    bs2.appendExistentialDimensions(bs1.nExist - bs2.nExist);
+  } else {
+    bs1.appendExistentialDimensions(bs2.nExist - bs1.nExist);
+  }
+
+  // TODO: Is there a better stratergy than this ? 
+  // Add the extra divisions from bs1 to bs2
+  unsigned extraDivs = bs1.divs.size() - bs2.divs.size();
+  bs2.insertDimensions(bs2.getNumDims() - 1, extraDivs);
+  for (unsigned i = 0; i < extraDivs; i++) {
+    DivisionConstraint &div = bs1.divs[bs2.getNumDivs()];
+    bs2.divs.push_back(DivisionConstraint(div.getCoeffs(), div.getDenominator(),
+                                          div.getVariable()));
+  }
+
+  // Loop over divs and find the equal ones
+  // This requires that divs in divs1 are ordered, which we ensured in
+  // simplify calls.
+  //
+  // Ordering ensures that divisions at index will not depend on any division
+  // at index >= i. This allows us to check same divisions easier, as well as
+  // Swaping will not cause any 
+  for (unsigned i = 0; i < bs2.divs.size(); i++) {
+    bool foundMatch = false;
+    for (unsigned j = i; j < bs1.divs.size(); j++) {
+      if (DivisionConstraint::sameDivision(bs1.divs[i], bs2.divs[j])) {
+        foundMatch = true;
+        if (i != j) {
+          bs2.swapDivisions(i, j);
+          break;
+        }
+      }
+    }
+
+    // Match not found, convert to existential
+    // Convert by swapping this division with this with the first division,
+    // removing this division, and incrementing nExist.
+    //
+    // This part leverages the order of variables in coefficients: Existentials,
+    // Divisions, Constant
+    if (not foundMatch) {
+      bs1.swapDivisions(bs1.getDivOffset(), i);
+      bs2.swapDivisions(bs2.getDivOffset(), i);
+
+      bs1.divs.erase(bs1.divs.begin());
```

Comment:
I think you should add the division ineqs here when you remove the divs

Reply:
Ill have to think about some things related to this. Ill add this as a TODO and get back to it later.